### PR TITLE
RUMM-3500 View tracking gap

### PR DESCRIPTION
**Pull request by xgouchet**
### What does this PR do?

Delay the view stopping event to prevent gaps between views

### Motivation

A non zero gap (~100ms) exists when navigating between fragments/activities. This is because a Fragment/Activity's `onResume` is called some time after the previous one's `onPause`. This would leave a short period of time during which errors and resources would not be tracked because no Active view was available. 

### Additional Notes

Thoughts about additional implementation of this feature : 
- mark a view as stopped as it is, and allow adding new children events (errors, resources), until a start view is detected or only after a specific period of time. 

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)

### CODE REVIEW

File: features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/domain/scope/RumViewManagerScope.kt [hunk 1]
```diff
@@ -59,6 +61,15 @@ internal class RumViewManagerScope(
 
         if (event is RumRawEvent.StartView && !stopped) {
             startForegroundView(event, writer)
+            lastStoppedViewTime?.let {
```
[the remark this answers is not in the document]
Comment (xgouchet) [reply features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/domain/scope/RumViewManagerScope.kt:64]: This would indeed, but the tracking strategy and view scope are completely independent from one another. If we were to move this detection in the view tracking strategy itself, we'd have to duplicate the logic three times. 
We do have the information of the view tracking strategy in the configuration telemetry event, meaning we can exclude orgs using manual tracking if need be to know where the gap is coming from.

File: features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/domain/scope/RumViewManagerScope.kt [hunk 2]
```diff
@@ -121,7 +132,13 @@ internal class RumViewManagerScope(
         val iterator = childrenScopes.iterator()
         @Suppress("UnsafeThirdPartyFunctionCall") // next/remove can't fail: we checked hasNext
         while (iterator.hasNext()) {
-            val result = iterator.next().handleEvent(event, writer)
+            val childScope = iterator.next()
+            if (event is RumRawEvent.StopView) {
```
Comment (mariusc83) [inline features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/domain/scope/RumViewManagerScope.kt:136]: shouldn't this be before the `while` ?

File: features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/tracking/AndroidXFragmentLifecycleCallbacks.kt [hunk 3]
```diff
@@ -77,11 +87,19 @@ internal open class AndroidXFragmentLifecycleCallbacks(
         }
     }
 
-    override fun onFragmentPaused(fm: FragmentManager, f: Fragment) {
-        super.onFragmentPaused(fm, f)
-        componentPredicate.runIfValid(f, internalLogger) {
-            val key = resolveKey(it)
-            rumMonitor.stopView(key)
+    @MainThread
+    override fun onFragmentStopped(fm: FragmentManager, f: Fragment) {
```
Comment (mariusc83) [inline features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/tracking/AndroidXFragmentLifecycleCallbacks.kt:91]: if we do this here, shouldn't we move the `startView` on the `onStart` also ? This will lower the gap.

File: features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/tracking/AndroidXFragmentLifecycleCallbacks.kt [hunk 4]
```diff
@@ -77,11 +87,19 @@ internal open class AndroidXFragmentLifecycleCallbacks(
...
+        super.onFragmentStopped(fm, f)
+        executor.scheduleSafe(
+            "Delayed view stop",
+            STOP_VIEW_DELAY_MS,
+            TimeUnit.MILLISECONDS,
+            sdkCore.internalLogger
+        ) {
+            componentPredicate.runIfValid(f, internalLogger) {
```
Comment (mariusc83) [inline features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/tracking/AndroidXFragmentLifecycleCallbacks.kt:99]: what happens if we delay the new view resumes/starts in before the 150ms interval ? In that case this scheduled action will not just stop it ?

File: features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/tracking/AndroidXFragmentLifecycleCallbacks.kt (hunk 3 above)
Comment (xgouchet) [reply features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/tracking/AndroidXFragmentLifecycleCallbacks.kt:91]: No, because the Fragment can be started but not visible yet (e.g.: in a view pager, the fragment is not resumed until visible but it can be started long before that)

File: features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/tracking/AndroidXFragmentLifecycleCallbacks.kt (hunk 4 above)
Comment (xgouchet) [reply features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/tracking/AndroidXFragmentLifecycleCallbacks.kt:99]: I'm not sure I understand the question you're asking. Why would we delay the new view resume/start?

Comment (mariusc83) [reply features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/tracking/AndroidXFragmentLifecycleCallbacks.kt:99]: let me rephrase this, if I am correct here you are postponing the `stopView` method call for the current view with 150ms. In case the next view `onResume` is called during this interval, your `stopView` will be called just after. I wanted to make sure in that case you are not going to stop the `just started new view`. The way I remember we are doing a validation on the event.key but I am not sure anymore that's the reason wanted to make sure you verified that.

File: features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/domain/scope/RumViewManagerScope.kt (hunk 2 above)
Comment (mariusc83) [reply features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/domain/scope/RumViewManagerScope.kt:136]: oh yeah I see that, nevermind my question.

File: features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/tracking/AndroidXFragmentLifecycleCallbacks.kt (hunk 4 above)
Comment (xgouchet) [reply features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/tracking/AndroidXFragmentLifecycleCallbacks.kt:99]: The stop view can only stop a view created with the same key, meaning we can't stop the newly created view.

File: instrumented/integration/src/androidTest/kotlin/com/datadog/android/sdk/integration/rum/FragmentTrackingTest.kt [hunk 5]
```diff
@@ -101,16 +101,16 @@ internal abstract class FragmentTrackingTest :
...
 
         // view stopped
-        expectedEvents.add(
-            ExpectedViewEvent(
-                fragmentAViewUrl,
-                3,
-                currentFragmentExtras(activity)
-            )
-        )
+//        expectedEvents.add(
```
Comment (xgouchet) [inline instrumented/integration/src/androidTest/kotlin/com/datadog/android/sdk/integration/rum/FragmentTrackingTest.kt:104]: This expected events is removed. Apparently the JUnit Runner's `callActivityOnStop` doesn't properly forward the fragment's `onStop()` to the fragment lifecycle listeners.

**Review by mariusc83 — APPROVED** [review]
LGTM

File: features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/domain/scope/RumViewManagerScope.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (plousada) [reply features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/domain/scope/RumViewManagerScope.kt:64]: Makes sense. 
It would be better if we could do it before sending the telemetry in the first place but detecting if we have an automated view tracking strategy in place here would create unwanted dependencies.

**Review by plousada — APPROVED** [review]

File: features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/tracking/AndroidXFragmentLifecycleCallbacks.kt [hunk 6]
```diff
@@ -94,4 +112,8 @@ internal open class AndroidXFragmentLifecycleCallbacks(
     }
 
     // endregion
+
+    companion object {
+        private const val STOP_VIEW_DELAY_MS = 200L
```
Comment (jonathanmos) [inline features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/tracking/AndroidXFragmentLifecycleCallbacks.kt:117]: I see that this constant is defined here but also in other places. Should we unify to a SSOT incase we want to modify the delay later on?

File: features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/tracking/OreoFragmentLifecycleCallbacks.kt [hunk 7]
```diff
@@ -104,6 +117,7 @@ internal class OreoFragmentLifecycleCallbacks(
 
     private companion object {
         private const val REPORT_FRAGMENT_NAME = "androidx.lifecycle.ReportFragment"
+        private const val STOP_VIEW_DELAY_MS = 200L
```
Comment (jonathanmos) [inline features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/tracking/OreoFragmentLifecycleCallbacks.kt:120]: see previous

File: features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/tracking/ActivityViewTrackingStrategy.kt [hunk 8]
```diff
@@ -84,4 +99,8 @@ class ActivityViewTrackingStrategy @JvmOverloads constructor(
     }
 
     // endregion
+
+    internal companion object {
+        private const val STOP_VIEW_DELAY_MS = 200L
```
Comment (jonathanmos) [inline features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/tracking/ActivityViewTrackingStrategy.kt:104]: see previous

**Review by jonathanmos — APPROVED** [review]
